Create z/tpf platform makefiles for building jit

pelemie opened the pull request:
The new makefiles allows a jit build for the z/TPF platform using gcc as
cross compiler.

Signed-off-by: pete lemieszewski <lemie@us.ibm.com>

DanHeidinga commented on runtime/tr.source/trj9/build/platform/host/s390-ztpf64-gcc.mk:
```diff
@@ -0,0 +1,26 @@
+# Copyright (c) 2000, 2017 IBM Corp. and others
```
Is this initial date correct?  The code here dates back to the 2000?

pelemie replied:
Thanks, since I've only just created the file, there's no need for two dates in the copyright statement.

DanHeidinga commented on runtime/tr.source/trj9/build/rules/ztpf-gcc/common.mk:
```diff
@@ -0,0 +1,72 @@
+# Copyright (c) 2000, 2017 IBM Corp. and others
```
Are the dates accurate here?

pelemie replied:
There should only be one copyright date.

DanHeidinga commented on runtime/tr.source/trj9/build/rules/ztpf-gcc/filetypes.mk:
```diff
@@ -0,0 +1,235 @@
+# Copyright (c) 2000, 2017 IBM Corp. and others
```
Is 2000 the correct start date?

pelemie replied:
There should only be one copyright date.

DanHeidinga commented on runtime/tr.source/trj9/build/rules/ztpf-gcc/filetypes.mk:
```diff
@@ -0,0 +1,235 @@
+
+RULE.s=$(eval $(DEF_RULE.s))
+
+##### START X SPECIFIC RULES #####
```
`START X SPECIFIC RULES` is this necessary in a ztpf file?

pelemie replied:
Not necessary.

DanHeidinga commented on runtime/tr.source/trj9/build/rules/ztpf-gcc/filetypes.mk:
```diff
@@ -0,0 +1,235 @@
+endif # ($(HOST_ARCH),x)
+##### END X SPECIFIC RULES #####
+
+##### START PPC SPECIFIC RULES #####
```
Is this section necessary in a ztpf file?

pelemie replied:
Not necessary...

DanHeidinga commented on runtime/tr.source/trj9/build/rules/ztpf-gcc/filetypes.mk:
```diff
@@ -0,0 +1,235 @@
+endif # ($(HOST_ARCH),z)
+##### END Z SPECIFIC RULES #####
+
+##### START ARM SPECIFIC RULES #####
```
Same question about this section - can it be removed?

DanHeidinga commented on runtime/tr.source/trj9/build/toolcfg/host/ztpf.mk:
```diff
@@ -0,0 +1,26 @@
+# Copyright (c) 2000, 2017 IBM Corp. and others
```
start date?

pelemie replied:
There should only be one copyright date.

pelemie commented:
@DanHeidinga the audit's copyright check fails after making the change to use (only) 2017. Here's the grep that fails  (I removed -q for quiet):
grep -E 'Copyright \(c\) **([0-9]{4}),** 2017 IBM Corp. and others' ztpf.mk  
Either the tool should upgrade the grep check or I could use "2017, 2017" in the copyright statement, regardless of how that looks. Thoughts?

pshipton commented:
Please use 2017, 2017

DanHeidinga commented:
Changes look ok to me.  Can someone more familiar with the JIT makefiles weigh in?  @0xdaryl can you suggest someone who can take a look?

mstoodle commented:
@fjeremic or @joransiu, could one of you please take a look here?

fjeremic commented on runtime/tr.source/trj9/z/env/J9CPU.cpp:
```diff
@@ -41,6 +41,7 @@
 
 #if defined(J9ZTPF)
 #include <tpf/c_pi1dt.h> // for access to machine type.
+#include <<tpf/c_cinfc.h> // for access to CINFC core tables.
```
Is this a typo? There seems to be an extra `<` in the path. Also why is this include needed exactly? There are no other code modifications that make use of the structures defined in this header.

pelemie replied:
Seems I cannot type; yes it is a typo. This include is needed to access a previously added field; copy/paste also is an issue for me.

joransiu commented on runtime/tr.source/trj9/z/env/J9CPU.cpp:
```diff
@@ -205,33 +206,17 @@ CPU::TO_PORTLIB_get390zLinuxMachineType()
+    * The z/TPF OS has strict requirements on supported z/Architecture hardware.
+    * As a result, the model number will necessarily match a valid TR_S390MachineType.
+    * Note - S390MachineTypes, declared above as a static const integer, contains the
+    * hex representation of the (decimal) model number found in the processors ID table.
```
Minor comment nit.  `S390MachineTypes` you described in an static const integer array containing the recognized machine types.   Are you referencing the `TR_S390MachineType` enum instead (https://github.com/eclipse/omr/blob/master/compiler/env/ProcessorInfo.hpp#L28-L61).  Also, not sure about the hex representation part... they are just binary integers. :)

pelemie replied:
The value that z/TPF picks up - the model number - matches TR_S390MachineType but not the binary integer. For example, model 2827 (ec12) was loaded into a register and compared against the value in the memory location for S390MachineTypes which is xB0B.  I can take the comment out but I wanted anyone on the Pok. team to be aware of why the checks are removed from this version of J9CPU.cpp

joransiu replied:
Hmm.. okay.   2827 == 0xB0B.... the representation in registers or memory should be identical in bits.  If you had left the checks in there, did you observe check failures?

pelemie replied:
@joransiu I updated the code to do what I should have done from the start... sprintf the data from core as character data and then convert it via atoi(). z/os kinda had to do the same thing although the data is already sprintf()'ed into uname. Then the checks work; otherwise they would fail.

joransiu commented on runtime/tr.source/trj9/build/toolcfg/ztpf-gcc/common.mk:
```diff
@@ -0,0 +1,300 @@
+# z/Architecture arch and tune level
+# z/TPF uses z10 as minminum supported z/Architecture.
+# mtune of z9-109 allows builtin versions of memcpy, etc.
+# instead of calling "arch-optimized" glibc functions.
```
This rational seems backwards to me -- and setting a tune to a level older than arch is odd.  Why would tune invoke arch-optimized glibc functions instead?  Are those compatible to run on z10 (min arch)?   

I would think you are leaving some performance on the table by not tuning to z196 or newer, as that level introduced the out of order pipeline.  However, I guess this gain may not be offset by the less efficient memcpy, etc?

joransiu replied:
OK.  Thanks for the extra insight.

pelemie commented:
@fjeremic and @joransiu, I have updated J9CPU.cpp again and we are using the version that I squashed yesterday. Please have a look. I think someone still needs to weigh in on the makefile updates as well. I understand today's a maintenance day and appreciate you taking the time to review this. Thanks.

joransiu commented:
@pelemie is there a new commit w/ changes we should be reviewing?   Have you made changes on top of https://github.com/eclipse/openj9/pull/234/commits/b0ad7f5646c3141b8d482d09c569d30569ef092c?

pelemie commented:
b0ad7f5 is the one I committed (and squashed) yesterday. J9CPU.cpp was changed in last night's commit, the mak files have remained the same.

joransiu commented:
@pelemie did you push last night's commit to your `submissionBranch`?  b0ad7f5 is the latest commit that shows up in that branch, and tracked by this PR.

pelemie commented:
@joransiu I did a rebase and another push/commit; there is a new ID associated with these changes so I think that is what you wanted?

fjeremic commented on runtime/tr.source/trj9/build/toolcfg/ztpf-gcc/common.mk:
```diff
@@ -0,0 +1,300 @@
+GENERATE_VERSION_SCRIPT?=$(JIT_SCRIPT_DIR)/generateVersion.pl
+
+# This is the script to preprocess ARM assembly files
+ARMASM_SCRIPT?=$(JIT_SCRIPT_DIR)/armasm2gas.sed
```
Is this needed?

pelemie replied:
no.

fjeremic commented on runtime/tr.source/trj9/z/env/J9CPU.cpp:
```diff
@@ -205,33 +206,17 @@ CPU::TO_PORTLIB_get390zLinuxMachineType()
-         }
-      }
-
+   TR_S390MachineType ret_machine = (TR_S390MachineType)(pid->pi1pids.pi1mslr.pi1mod);
```
Are you certain `pid->pi1pids.pi1mslr.pi1mod` will always contain a valid processor ID? What happens if this value is not contained within the list of enumerations in `TR_S390MachineType`? Then the value of `ret_machine` is technically undefined. Will this break things?

pelemie replied:
While we get the data at ipl time from the store cpuid instruction (STIDP)  there is no reason to think that this area in main storage couldn't be corrupted at any time.  I will reset the code to do what z/os essentially does; get the character representation of the model, convert it to integer via atoi and then test the integrity of the result.

fjeremic replied:
Or simply iterate through the list of machines we support like the code did previously:

```
-   // Scan list of supported machines.
-   for (int i = 0; i < sizeof(S390MachineTypes) / sizeof(int); ++i)
-      {
-      if (machine == S390MachineTypes[i])
-         {
-         ret_machine = (TR_S390MachineType)machine;
-         }
-      }
```

pelemie commented:
Thanks @fjeremic I put the code back into J9CPU and had to sprintf/atoi the cpu model number to get things to work correctly.

fjeremic commented:
Jenkins test sanity

fjeremic commented:
@joransiu I'll delegate merging to you if you have no additional concerns regarding this PR.
